Question1

stkhrmwl opened the pull request:
課題
---
[課題1](https://github.com/Caraquri/intern_2week_study_ios/issues/1)

概要
---
1. addButtonを押した際にtextFieldに入っている文字をtextViewに追加する処理を次の条件を満たして実装
    - 起動時にtextViewを空にする
    - textViewは改行してから追加する
2. clearTextButtonを追加し，textViewの文字を消す処理を実装
3. addButtonを押した際，textFieldに文字が入っていない場合はtextViewに追加できないようにする処理を追加．このときtextFieldの下にLabelを追加し「赤文字」で「文字を入力してください」と警告を表示する．

特に確認して頂きたい項目
---
- Question1ViewController.swiftの`addText(_ sender: Any)`関数での`if let`の使い方

動作確認用 Screenshots
---
![スクリーンショット 2020-08-12 15 45 50](https://user-images.githubusercontent.com/64292043/89983857-5dd94c00-dcb3-11ea-8279-7660aff50bba.png)
Clearボタンはキーボードに隠れてます
![スクリーンショット 2020-08-12 15 45 57](https://user-images.githubusercontent.com/64292043/89983908-76496680-dcb3-11ea-8658-5316e3a6318d.png)

参考URL
---
- [guard letとif letの使い分け - Qiita](https://qiita.com/solty_919/items/8b22c8b30cd1a5596c83)
- [swiftでテキストフィールドが空かどうかを確認する方法](https://www.it-swarm.dev/ja/swift/swiftでテキストフィールドが空かどうかを確認する方法/1047233533/)
- [[Tips] UILabel, UITextField, UITextViewのちょっとした挙動の違い - Qiita](https://qiita.com/ktanaka117/items/98c9aef9e8c15a92ed56)
- [ボタンのタッチイベント(クリックイベント)の実装 - ボタンのタップ時に処理を実行する - Swift プログラミング](https://www.ipentec.com/document/mac-os-x-swift-button-touch-down-event-implementation)
- [Swiftの関数の基本的な取り扱い](https://thinkit.co.jp/article/14301)

stkhrmwl committed:
起動時にtextViewを空にする

stkhrmwl committed:
Addを押した際にtextFieldに入っている文字をtextViewに改行して追加

stkhrmwl committed:
Clearボタンを追加 (置いただけ)

stkhrmwl committed:
ClearボタンをAutoLayoutで配置

stkhrmwl committed:
clearTextButtonを追加し，textViewの文字を消す処理を追加

stkhrmwl committed:
textFieldに文字が入っていない場合はtextViewに追加できないようにする処理を追加

stkhrmwl committed:
Labelを追加 (置いただけ)

stkhrmwl committed:
LabelをAutoLayoutで配置

stkhrmwl committed:
文字が入力されずAddが押された際に警告を表示

stkhrmwl committed:
getValueOnTextField()の返り値を非アンラップに変更
addTextのif文をif letに変更

stkhrmwl committed:
textFieldの外がタップされた際にキーボードを閉じる処理を追加
returnがタップされた際にキーボードを閉じる処理を追加

stkhrmwl committed:
storyboardの制約のズレを修正

msugai-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -1,12 +1,65 @@
 import UIKit
 
-final class Question1ViewController: UIViewController {
+final class Question1ViewController: UIViewController, UITextFieldDelegate {
```
長くなるのでdelegate などはExtensionを使って、区切ります。
```
final class Question1ViewController: UIViewController{

}

Question1ViewController: UITextFieldDelegate {
// UITextFieldDelegateのイベント
}
```

stkhrmwl replied:
https://github.com/stkhrmwl/caraquri-intern-2weeks/blob/0fa24698d8ef49ffbf378667c5d6d98d657f271f/intern_2week_study/Question/Question1/Question1ViewController.swift#L37-L43

クラス外で拡張しました

msugai-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -1,12 +1,65 @@
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        self.setValueOnTextView(value: "")
```
全体的にself不要です。
本来不要である箇所につける必要はないです。
必要な際のみつけて下さい。

msugai-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -1,12 +1,65 @@
+    }
+
+    // addTextButtonイベント
+    @IBAction func addText(_ sender: Any) {
```
swiftだと基本的にエラー処理はguard let を使って下さい。

基本的に なるべく否定せずに書きましょう(他の人が読みやすいようにしましょう)
```
!addStr.isEmpty 
```

msugai-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -1,12 +1,65 @@
+    }
+    
+    // clearTextButtonイベント
+    @IBAction func clearText(_ sender: Any) {
```
Swiftだとメソッドを使う場合は動詞で始まりますね、コメントせずにわかる名前にしましょう
```
didClearButtonTouched()
```

msugai-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -1,12 +1,65 @@
+    }
+    
+    // labelに文字列をセットする
+    private func setValueOnLabel(value: String) {
```
メソッドを使うなら
```
private func setValueOnLabel(_ value: String) {
}
```
のように _をつけてあげると
```
 setValueOnLabel("test") 
```
のようにすっきり書ける。

1行だけの処理ならそもそもメソッドを使わなくても良いですね

stkhrmwl committed:
ボタンのUIを日本語に変更

stkhrmwl committed:
UITextFieldDelegateの継承をextensionとしてclass外に定義

stkhrmwl committed:
メンバ変数の余分なselfを除去
処理が少ない関数を除去

stkhrmwl committed:
ユーザ目線の関数名に変更 (例: addText -> didAddButtonTouched)

stkhrmwl committed:
if letをguard letに変更

stkhrmwl committed:
次の記事に従って関数名をUIと関連しない単純なものに変更

なぜ IBAction に didTapButton という名前をつけてはいけないのか - Qiita
https://qiita.com/bitpoetics/items/a63bd1d9afc0b3221890

stkhrmwl committed:
ボタンのインスタンス名を変更
